order implementation

ADIGrimm opened the pull request.

ADIGrimm committed:
order implementation

Elena-Bruyako commented on src/main/java/online/bookstore/controller/OrderController.java:
```diff
@@ -0,0 +1,49 @@
+            description = "Return list of all orders of user")
+    @GetMapping
+    @PreAuthorize("hasAnyRole('ROLE_USER')")
+    public Page<OrderDto> getAll(Authentication authentication, Pageable pageable) {
```
you can add
```suggestion
    public Page<OrderDto> getAll(Authentication authentication, @ParameterObject @PageableDefault Pageable pageable) {
```

Elena-Bruyako commented on src/main/java/online/bookstore/controller/OrderController.java:
```diff
@@ -0,0 +1,49 @@
+            description = "Create order")
+    @PostMapping
+    @PreAuthorize("hasAnyRole('ROLE_USER')")
+    public OrderDto createOrder(Authentication authentication, @RequestBody String address) {
```
It's better to create Dto

```suggestion
    public OrderDto createOrder(Authentication authentication, @RequestBody @Valid CreateOrderRequestDto orderDto) {
```

Elena-Bruyako commented on src/main/java/online/bookstore/controller/OrderController.java:
```diff
@@ -0,0 +1,49 @@
+            "Updates the order status to one of these: delivered, cancelled, pending")
+    @PutMapping("/{id}")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN')")
+    public OrderDto updateStatus(@PathVariable Long orderId, @RequestBody Order.Status status) {
```
same here

Elena-Bruyako commented on src/main/java/online/bookstore/controller/OrderItemController.java:
```diff
@@ -0,0 +1,40 @@
+@RequestMapping("/orders/{orderId}/items")
+@RequiredArgsConstructor
+@RestController
+public class OrderItemController implements UserContextHelper {
```
remove controller, all endpoints should be in OrderController

Elena-Bruyako commented on src/main/java/online/bookstore/model/Order.java:
```diff
@@ -0,0 +1,53 @@
+    private Set<OrderItem> orderItems = new HashSet<>();
+
+    public enum Status {
+        Pending,
```
use upper case only for enum

Elena-Bruyako commented on src/main/java/online/bookstore/model/OrderItem.java:
```diff
@@ -0,0 +1,34 @@
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+    @OneToOne
```
Add lazy and joincolumn

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/OrderItemServiceImpl.java:
```diff
@@ -0,0 +1,29 @@
+
+@Service
+@RequiredArgsConstructor
+public class OrderItemServiceImpl implements OrderItemService {
```
remove, replace logic to OrderService

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+        return orderRepository.findAllByUserId(userId, pageable).map(orderMapper::toDto);
+    }
+
+    @Override
```
think where is Transactional needed

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+        order.setUser(userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find user by id" + userId)));
+        order.setStatus(Order.Status.Pending);
+        order.setOrderItems(shoppingCartRepository.findById(userId).get().getCartItems().stream()
```
shop cart could be empty
improve readability

Elena-Bruyako reviewed (changes requested):
see cooments

ADIGrimm committed:
fixed

added @ParameterObject @PageableDefault annotations
OrderItem removed into Order
created CreateOrderRequestDto and UpdateOrderStatusDto
changed create() in OrderServiceImpl
enum variants in Order now uppercase
fixed annotations for field book in OrderItem

Elena-Bruyako commented on src/main/java/online/bookstore/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+        order.setStatus(Order.Status.PENDING);
+        Set<CartItem> cartItems = shoppingCartRepository.findById(userId).get().getCartItems();
+        if (cartItems.isEmpty()) {
+            throw new RuntimeException("Shopping cart is empty: " + cartItems);
```
```suggestion
            throw new OrderProcessingException("Shopping cart is empty for user " + userId);
```

Elena-Bruyako reviewed (changes requested):
one comment

ADIGrimm committed:
fixed

created OrderProcessingException
changed throw exception in create() in OrderServiceImpl

ahoienko reviewed (approved):
Good job!
